File: src/main/java/no/fintlabs/model/instance/InstanceObject.java
package no.fintlabs.model.instance;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.extern.jackson.Jacksonized;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;


@Getter
@Builder
@EqualsAndHashCode
@Jacksonized
public class InstanceObject {

    @Builder.Default
    private Map<String, String> valuePerKey = new HashMap<>();

    @Builder.Default
    private Map<String, Collection<InstanceObject>> objectCollectionPerKey = new HashMap<>();

}

File: src/main/kotlin/no/novari/flyt/mapping/model/instance/InstanceObject.kt
package no.novari.flyt.mapping.model.instance

data class InstanceObject(
    val valuePerKey: Map<String, String?> = emptyMap(),
    val objectCollectionPerKey: Map<String, Collection<InstanceObject>> = emptyMap(),
) {
    class Builder {
        private var valuePerKey: Map<String, String?>? = null
        private var objectCollectionPerKey: Map<String, Collection<InstanceObject>>? = null

        fun valuePerKey(valuePerKey: Map<String, String?>) = apply { this.valuePerKey = valuePerKey }

        fun objectCollectionPerKey(objectCollectionPerKey: Map<String, Collection<InstanceObject>>) =
            apply { this.objectCollectionPerKey = objectCollectionPerKey }

        fun build(): InstanceObject {
            return InstanceObject(
                valuePerKey = valuePerKey ?: emptyMap(),
                objectCollectionPerKey = objectCollectionPerKey ?: emptyMap(),
            )
        }
    }

    companion object {
        @JvmStatic
        fun builder(): Builder = Builder()
    }
}
